### PR TITLE
bsunanda:Run2-hcx122 Testing Reconstruction geometry for Plan1 option

### PR DESCRIPTION
**Pull request by bsunanda**
Tests geometry of merged cells match with the initial cells in Plan 1 operation of HCAL.

### CODE REVIEW

**Comment by bsunanda** [comment]
@cmsbuild Please test

**Comment by bsunanda** [comment]
@cmsbuild Please test

**Comment by bsunanda** [comment]
@smuzaffar @davidlange6 The failure has nothing to do with this PR.

**Comment by bsunanda** [comment]
@cmsbuild Please test

File: Geometry/HcalCommonData/src/HcalDDDRecConstants.cc [hunk 1]
```diff
@@ -595,6 +595,20 @@ void HcalDDDRecConstants::specialRBXHBHE(const std::vector<HcalDetId>& idsOld,
   }
 }
 
+bool HcalDDDRecConstants::specialRBXHBHE(bool tobemerged,
+					 std::vector<HcalDetId>& ids) const {
+  ids.clear();
```
Comment (ianna) [inline Geometry/HcalCommonData/src/HcalDDDRecConstants.cc]: @bsunanda - why do you need to clear the vector here? You fill an empty newly created vector from the way it is used in the test. Note, that the capacity (memory usage) of a vector will never decrease, even if you call the clear method, unless you use the swap method.

**Review by ianna — CHANGES_REQUESTED** [review]

**Comment by ianna** [comment]
@bsunanda - please, provide description

**Comment by bsunanda** [comment]
@cmsbuild Please test

File: Geometry/HcalTowerAlgo/test/HcalGeometryPlan1Tester.cc [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+				      const edm::EventSetup& iSetup) {
+
+  edm::ESHandle<HcalDDDRecConstants> hDRCons;
+  iSetup.get<HcalRecNumberingRecord>().get(hDRCons);
+  const HcalDDDRecConstants hcons = (*hDRCons);
+  edm::ESHandle<HcalTopology> topologyHandle;
+  iSetup.get<HcalRecNumberingRecord>().get(topologyHandle);
+  const HcalTopology topology = (*topologyHandle);
+  HcalGeometry* geom(0);
+  if (geomDB_) {
```
Comment (ianna) [inline Geometry/HcalTowerAlgo/test/HcalGeometryPlan1Tester.cc]: @bsunanda - the geometry will not necessarily come from DB, but from an EventSetup. The python configuration will define if it is coming from DB.

File: Geometry/HcalTowerAlgo/test/HcalGeometryPlan1Tester.cc [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+    edm::ESHandle<CaloGeometry> pG;
+    iSetup.get<CaloGeometryRecord>().get(pG);
+    const CaloGeometry* geo = pG.product();
+    geom = (HcalGeometry*)(geo->getSubdetectorGeometry(DetId::Hcal,HcalBarrel));
+  } else {
+    HcalFlexiHardcodeGeometryLoader m_loader(ps0_);
```
Comment (ianna) [inline Geometry/HcalTowerAlgo/test/HcalGeometryPlan1Tester.cc]: @bsunanda - a possibility of an HcalGeometry creation outside of an EventSetup is a bad practice. I hope, it is limited to the test only.

File: Geometry/HcalTowerAlgo/test/runHcalGeometryPlan1Tester_cfg.py [hunk 4]
```diff
@@ -0,0 +1,17 @@
+import FWCore.ParameterSet.Config as cms
+
+process = cms.Process("HcalGeometryTest")
+
+process.load("Geometry.HcalCommonData.testGeometry17bXML_cfi")
```
Comment (ianna) [inline Geometry/HcalTowerAlgo/test/runHcalGeometryPlan1Tester_cfg.py]: @bsunanda - here is where the geometry choice is made and it is from XML, not DB

File: Geometry/HcalTowerAlgo/test/HcalGeometryPlan1Tester.cc [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+    geom = (HcalGeometry*)(m_loader.load(topology, hcons));
+  }
+
+  std::vector<HcalDetId> idsp;
+  bool ok = hcons.specialRBXHBHE(true,idsp);
+  std::cout << "Special RBX Flag " << ok << " with " << idsp.size()
+	    << " ID's" << std::endl;
+  int nall(0), ngood(0);
+  for (unsigned int k=0; k<idsp.size(); ++k) {
```
Comment (ianna) [inline Geometry/HcalTowerAlgo/test/HcalGeometryPlan1Tester.cc]: @bsunanda - please, use a range-based for loop

File: Geometry/HcalCommonData/src/HcalDDDRecConstants.cc [hunk 6]
```diff
@@ -595,6 +595,19 @@ void HcalDDDRecConstants::specialRBXHBHE(const std::vector<HcalDetId>& idsOld,
   }
 }
 
+bool HcalDDDRecConstants::specialRBXHBHE(bool tobemerged,
+					 std::vector<HcalDetId>& ids) const {
+  if (tobemerged) {
+    std::map<HcalDetId,HcalDetId>::const_iterator itr;
```
Comment (ianna) [inline Geometry/HcalCommonData/src/HcalDDDRecConstants.cc]: @bsunanda - please, consider using range-based for loops

**Review by ianna — CHANGES_REQUESTED** [review]

**Comment by ianna** [comment]
@bsunanda - please, do not put neither branch name nor a PR number in the a PR title. The title is used in release notes, for example:
You can have a look https://github.com/cms-sw/cmssw/releases/CMSSW_9_0_0
If you need to refer to another PR, it is better to do it in the PR description - a link to it will be automatically added.

**Comment by bsunanda** [comment]
@cmsbuild Please test

File: Geometry/HcalTowerAlgo/test/HcalGeometryPlan1Tester.cc [hunk 7]
```diff
@@ -57,16 +57,17 @@ void HcalGeometryPlan1Tester::analyze(const edm::Event& /*iEvent*/,
   std::cout << "Special RBX Flag " << ok << " with " << idsp.size()
 	    << " ID's" << std::endl;
   int nall(0), ngood(0);
-  for (unsigned int k=0; k<idsp.size(); ++k) {
-    if (topology.valid(idsp[k])) {
+  for (std::vector<HcalDetId>::const_iterator itr=idsp.begin();
```
Comment (ianna) [inline Geometry/HcalTowerAlgo/test/HcalGeometryPlan1Tester.cc]: @bsunanda - the following would be shorter. Please, have a look at http://en.cppreference.com/w/cpp/language/range-for
```
for( const auto& itr : idsp )
{
 ...
} 
```

**Comment by ianna** [comment]
+1

**Comment by davidlange6** [comment]
+1

**Pull request merged**